### ISSUE TITLE
Failed to update Foxx store from GitHub

### ISSUE THREAD

**Issue by santo74** [arangodb/arangodb#9121]
## My Environment

* __ArangoDB Version__: 3.4.5
* __Storage Engine__: RocksDB
* __Deployment Mode__: Single Server
* __Deployment Strategy__: ArangoDB Starter
* __Configuration__:
* __Infrastructure__: Openstack
* __Operating System__: Ubuntu 16.04
* __Total RAM in your machine__: 8GB
* __Disks in use__: SSD
* __Used Package__: TAR package

## Component, Query & Data

__Affected feature__:
General operation, Foxx related

## Steps to reproduce

__Problem__:
(Almost) daily I see the following warning in the logs:

> Failed to update Foxx store from GitHub.

Why should the "Foxx store" be updated from GitHub and if so, then why is it failing?
I'm not using any Foxx applications myself, it's just a stock ArangoDB install with 2 databases

__Expected result__:
I see no reason why Foxx should be updated from GitHub, but if there is, then it shouldn't fail.

**Comment by OmarAyo** [comment]
Hi @santo74 

>Why should the "Foxx store" be updated from GitHub

The Foxx store will fetch the apps directly from GitHub, this is how is designed 

> why is it failing?

Probably your instance is not connected to the internet or probably you have some network issue. The Foxx store will fetch for apps directly from GitHub even if you are not using any Foxx app 

On the other hand, those messages can be ignored, if you are not using or willing to use any Foxx app taken from the Github store 

Best,

**Comment by santo74** [comment]
@OmarAyo thanks for the clarification.

However, I think this shouldn't be on by default.
By default no Foxx apps are enabled, so there is no need to fetch from GitHub.
Security wise this isn't good practice in my opinion.

Btw, on that server I'm not blocking any outgoing connections for the moment.

**Comment by OmarAyo** [comment]
Hi @santo74 

I will forward your feedback internally, thanks for the feedback. 

Regarding the root cause of those messages, if you are not blocking outgoing connections, then probably your firewall or proxy is disallowing access to GitHub.

Best,

**Comment by santo74** [comment]
> I will forward your feedback internally, thanks for the feedback.

Thanks!

> Regarding the root cause of those messages, if you are not blocking outgoing connections, then probably your firewall or proxy is disallowing access to GitHub.

I have no knowledge of GitHub being blocked/disallowed.
For my own projects I am using GitLab and I can fetch packages from there without any issues.

**Comment by dothebart** [comment]
Fetching the service list from github seems to be broken in 3.4 and devel.

**Comment by santo74** [comment]
@dothebart thanks for investigating this further and confirming the issue.
When fixing this, would it be possible to make the fetching "optional", i.e. only fetch when Foxx apps are effectively in use (as I described earlier)?

**Comment by jsteemann** [comment]
@dothebart: I can retrieve the list of Foxx apps using 3.4 and devel just fine, and can even install them without problems. 
I have been using a local ArangoDB installation and am not running it inside a specific container or such.

@santo74 : starting with ArangoDB 3.5, fetching the apps from the store can be turned off using a configuration option.

**Comment by hkernbach** [comment]
> Fetching the service list from github seems to be broken in 3.4 and devel.

@dothebart is there more information available?

**Comment by santo74** [comment]
> @dothebart: I can retrieve the list of Foxx apps using 3.4 and devel just fine, and can even install them without problems.
> I have been using a local ArangoDB installation and am not running it inside a specific container or such.
> 
> @santo74 : starting with ArangoDB 3.5, fetching the apps from the store can be turned off using a configuration option.

@jsteemann that's great! Any chance it will be backported to 3.4?

**Comment by jsteemann** [comment]
@santo74: what I have been trying to say in my last message is that it worked for me out of the box. 
I didn't fix anything, and it would be good if @dothebart could back his claims with some info about the environment he ran it in.

**Comment by hkernbach** [comment]
@jsteemann @dothebart's issue is not related to this ticket here.

**Comment by santo74** [comment]
> @santo74: what I have been trying to say in my last message is that it worked for me out of the box.
> I didn't fix anything, and it would be good if @dothebart could back his claims with some info about the environment he ran it in.

@jsteemann I know, but you said actually 2 things ;-)
So what I was trying to say is whether the possibility to turn off fetching the apps might be backported to 3.4?

**Comment by dothebart** [comment]
@santo74 this is [a new feature](https://www.arangodb.com/docs/3.5/security-security-options.html) and it won't be backported. 

Regarding the sole problem of this issue, I tried to download a similar file as arangodb desires from inside my arangodb docker container like this: 

Find the container with arangod:
```
docker ps 
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                    NAMES
73c5dc966862        arangodb:3.4.5      "/entrypoint.sh ar..."   14 seconds ago      Up 13 seconds       0.0.0.0:8529->8529/tcp   happy_borg
```
then execute an interactive shell in it, and use arangosh to try to download something from github:
```
docker exec -it 73c5dc966862 /bin/sh
/usr/bin/arangosh
 x=require('internal').download('https://raw.githubusercontent.com/arangodb-foxx/tableau-connector/v1.0.0/manifest.json')
```
This works for me like this:
```
{ 
  "headers" : { 
    "source-age" : "0", 
    "expires" : "Mon, 03 Jun 2019 12:06:04 GMT", 
    "x-fastly-request-id" : "fdc6a0943ea261159fab77cd77a5bc4d7d143da5", 
    "etag" : "\"be2919687748687f0703e68e44862dc5c82c3e46\"", 
    "x-xss-protection" : "1; mode=block", 
    "x-frame-options" : "deny", 
    "content-type" : "text/plain; charset=utf-8", 
    "connection" : "keep-alive", 
    "x-geo-block-list" : "", 
    "x-served-by" : "cache-hhn1531-HHN", 
    "strict-transport-security" : "max-age=31536000", 
    "content-length" : "383", 
    "accept-ranges" : "bytes", 
    "content-security-policy" : "default-src 'none'; style-src 'unsafe-inline'; sandbox", 
    "x-content-type-options" : "nosniff", 
    "date" : "Mon, 03 Jun 2019 12:01:04 GMT", 
    "http/1.1" : "200 OK", 
    "via" : "1.1 varnish", 
    "x-github-request-id" : "3246:3591:546780:63A572:5CF50C00", 
    "x-timer" : "S1559563265.866430,VS0,VE89", 
    "cache-control" : "max-age=300", 
    "x-cache" : "MISS", 
    "x-cache-hits" : "0", 
    "vary" : "Authorization,Accept-Encoding", 
    "access-control-allow-origin" : "*" 
  }, 
  "body" : "{\n  \"$schema\": \"http://json.schemastore.org/foxx-manifest\",\n  \"name\": \"tableau-connector\",\n  \"description\": \"Example Tableau Web Data Connector for ArangoDB.\",\n  \"version\": \"1.0.0\",\n  \"thumbnail\": \"tableau.png\",\n  \"engines\": {\n    \"arangodb\": \"^3.0.0\"\n  },...", 
  "code" : 200, 
  "message" : "OK" 
}
```
Can you confirm that it does or does not work for you too?

You're naming openshift as environment, which is known to have stricter permissions than docker - this could be a reason.

**Comment by santo74** [comment]
> @santo74 this is [a new feature](https://www.arangodb.com/docs/3.5/security-security-options.html) and it won't be backported.

Ok, fair enough.

> Regarding the sole problem of this issue, I tried to download a similar file as arangodb desires from inside my arangodb docker container like this:
> 
> Find the container with arangod:
> 
> ```
> docker ps 
> CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                    NAMES
> 73c5dc966862        arangodb:3.4.5      "/entrypoint.sh ar..."   14 seconds ago      Up 13 seconds       0.0.0.0:8529->8529/tcp   happy_borg
> ```
> 
> then execute an interactive shell in it, and use arangosh to try to download something from github:
> 
> ```
> docker exec -it 73c5dc966862 /bin/sh
> /usr/bin/arangosh
>  x=require('internal').download('https://raw.githubusercontent.com/arangodb-foxx/tableau-connector/v1.0.0/manifest.json')
> ```
> 
> This works for me like this:
> 
> ```
> { 
>   "headers" : { 
>     "source-age" : "0", 
>     "expires" : "Mon, 03 Jun 2019 12:06:04 GMT", 
>     "x-fastly-request-id" : "fdc6a0943ea261159fab77cd77a5bc4d7d143da5", 
>     "etag" : "\"be2919687748687f0703e68e44862dc5c82c3e46\"", 
>     "x-xss-protection" : "1; mode=block", 
>     "x-frame-options" : "deny", 
>     "content-type" : "text/plain; charset=utf-8", 
>     "connection" : "keep-alive", 
>     "x-geo-block-list" : "", 
>     "x-served-by" : "cache-hhn1531-HHN", 
>     "strict-transport-security" : "max-age=31536000", 
>     "content-length" : "383", 
>     "accept-ranges" : "bytes", 
>     "content-security-policy" : "default-src 'none'; style-src 'unsafe-inline'; sandbox", 
>     "x-content-type-options" : "nosniff", 
>     "date" : "Mon, 03 Jun 2019 12:01:04 GMT", 
>     "http/1.1" : "200 OK", 
>     "via" : "1.1 varnish", 
>     "x-github-request-id" : "3246:3591:546780:63A572:5CF50C00", 
>     "x-timer" : "S1559563265.866430,VS0,VE89", 
>     "cache-control" : "max-age=300", 
>     "x-cache" : "MISS", 
>     "x-cache-hits" : "0", 
>     "vary" : "Authorization,Accept-Encoding", 
>     "access-control-allow-origin" : "*" 
>   }, 
>   "body" : "{\n  \"$schema\": \"http://json.schemastore.org/foxx-manifest\",\n  \"name\": \"tableau-connector\",\n  \"description\": \"Example Tableau Web Data Connector for ArangoDB.\",\n  \"version\": \"1.0.0\",\n  \"thumbnail\": \"tableau.png\",\n  \"engines\": {\n    \"arangodb\": \"^3.0.0\"\n  },...", 
>   "code" : 200, 
>   "message" : "OK" 
> }
> ```
> 
> Can you confirm that it does or does not work for you too?

I'm not using docker, so executed it directly from arangosh on the server and it worked fine.
Therefore I also rechecked the arangodb logs and noticed that the warning hasn't appeared anymore since I created this issue.
Can it be that the fetching is only performed during server startup?
Last week I restarted ArangoDB (or the host server it runs on) several times, but since I created this issue it wasn't restarted anymore.

> You're naming openshift as environment, which is known to have stricter permissions than docker - this could be a reason.

No, the environment where ArangoDB runs is with a cloud provider whose infrastructure is based on OpenStack, which is not the same as OpenShift.

**Comment by dothebart** [comment]
Ok, maybe at the time you were trying this github was having issues? (yes, that may also happen ;-) )

**Comment by santo74** [comment]
Yes that can happen indeed :-)
However, it happened at least 3 days in a row and I'm not aware of any recent "major" connectivity issues with GitHub.

I suggest that I monitor the logs for the next couple of days and if it doesn't happen again, I'll close the issue because in that case it can't be reproduced (anymore).

**Comment by santo74** [comment]
It appears to happen very randomly.
Right after my last comment it happened 2 times in 1 day, then everything was fine until yesterday.
I attached a screenshot to give an idea of the randomness.

![Screenshot_2019-06-11 ArangoDB Web Interface](https://user-images.githubusercontent.com/7080464/59265270-5056ab00-8c45-11e9-9171-ab71a5da74a9.png)

**Comment by dothebart** [comment]
@pluma can we add the details of the exception to this message? 
```router.get('/fishbowl', function (req, res) {
  if (internal.isFoxxStoreDisabled()) {
    res.json([]);
  } else {
    try {
      store.update();
    } catch (e) {
      console.warn('Failed to update Foxx store from GitHub.');
    }
    res.json(store.availableJson());
  } 
})
.summary('List of all Foxx services submitted to the Foxx store.')
.description(dd`
  This function contacts the fishbowl and reports which services are available for install.
`);
```

**Comment by pluma** [comment]
We could but I doubt the exception will be significantly more useful. We could log information about the response as that could indicate at least whether it failed because of an error response or because of network issues.

**Comment by dothebart** [comment]
shouldn't this be an exception from the HTTP connection or its flow? so you could see whether arangod wasn't able to connect, or whether github itself errored out.

**Comment by Simran-B** [comment]
> By default no Foxx apps are enabled, so there is no need to fetch from GitHub.

BTW: There are in fact two system services, `/_admin/aardvark` and `_api/foxx`.

**Comment by pluma** [comment]
@Simran-B True but the only thing that fetches from GitHub is the Foxx Store, which is only accessible via the Aardvark web interface. Note that if the Foxx Store is enabled, Aardvark will trigger a fetch as soon as you click on the Foxx services tab because the "install from store" view is the default.

**Comment by Simran-B** [comment]
FYI: The option in 3.5 is called `--foxx.store` and true by default.

**Comment by santo74** [comment]
Thanks for the follow-up guys!
Providing an indication of why the store update fails might be a good idea indeed.

It's also great to know the store can be disabled in 3.5, but that version isn't released yet and I probably won't be able to upgrade to 3.5 anytime soon after the release (for several reasons).
In production we're trying to skip any .0 and even .1 releases whenever possible.
Also, if I'm not mistaken "simple queries" will be removed in 3.5 but at the moment several of our services still rely on "simple queries". This means part of our code has to be rewritten in order to be able to upgrade to 3.5.

**Comment by Simran-B** [comment]
The Simple Queries API and interfaces are marked for deprecation, but will still be available until v4.0.0.

**Comment by maxkernbach** [comment]
Disabling the Foxx store has been backported to the 3.4 branch by https://github.com/arangodb/arangodb/pull/9593.

> 
> * Added options to make server more secure:
>   - `--foxx.api`: set to false disables Foxx management API
>   - `--foxx.store`: set to false disables Foxx UI

The new options will be available in 3.4.8.

**Comment by santo74** [comment]
@maxkernbach thanks for backporting this to 3.4! 
I guess this issue can be closed now.

**Comment by maxkernbach** [comment]
@santo74 Sure, we will close this issue once version 3.4.8 is released.

**Comment by maxkernbach** [comment]
Closing this issue as fixed since version 3.4.8, which supports the following options, has been released:

- `--foxx.api`: set to false disables Foxx management API
- `--foxx.store`: set to false disables Foxx UI

Please find the latest version in the [download section](https://www.arangodb.com/download-major/).
